src/main/java/com/epam/wl/dao/BookDAO.java

```diff
@@ -0,0 +1,143 @@
+package com.epam.wl.dao;
+
+import com.epam.wl.entities.Book;
+import com.epam.wl.entities.BookInstance;
+import com.sun.org.apache.regexp.internal.RE;
+
+import javax.sql.DataSource;
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+public class BookDAO {
+    private DataSource dataSource;
+
+    public BookDAO(DataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    public List<Book> getAllBooks() {
+        List<Book> books = new ArrayList<>();
+
+        try (Connection connection = dataSource.getConnection()) {
+            final Statement statement = connection.createStatement();
+            final ResultSet resultSet = statement.executeQuery("SELECT * FROM book");
+
+            while (resultSet.next()) {
+                books.add(new Book(resultSet.getInt("id"), resultSet.getString("author"),
+                        resultSet.getString("title"), resultSet.getInt("year")));
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+
+        return books;
+    }
+
+    public List<BookInstance> getAllBookInstances() {
+        List<BookInstance> bookInstances = new ArrayList<>();
+
+        try (Connection connection = dataSource.getConnection()) {
+            final Statement statement = connection.createStatement();
+            final ResultSet resultSet = statement.executeQuery("SELECT * FROM book_instance");
+
+            while (resultSet.next()) {
+                bookInstances.add(new BookInstance(resultSet.getInt("id"),
+                        resultSet.getInt("bookid")));
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+
+        return bookInstances;
+    }
+
+    public int getBookId(String author, String title, int year) {
+        int id = -1;
```

Comment:
This is not a good practice. I understand that you're using this to get around the possibility of a non-existent ID in a database.If you want to protect your application from it, you can just throw an exception in the if block below, without using template magic numbers. On the other hand, I see that you're returning -1 if the book does not exist. While this is not the most elegant solution, it's still acceptable, but you can just return -1 in the if block, without creating a seemingly useless int.

Reply:
I think this can be changed in every method in this file.

Reply:
Thanks for the detailed comment, I'll fix it.